File: QueueGUI.py
import tkinter as tk
from tkinter import ttk, messagebox
from collections import deque
import pyttsx3
import threading
import platform

class QueueGUI:
    def __init__(self, root):
        self.root = root
        self.queue = deque()
        self.counter = 0
        
        # Warna tema Beige (sama persis dengan Java)
        self.beige_bg = "#FAF5EB"      # Color(250, 245, 235)
        self.beige_panel = "#F5F0E4"    # Color(245, 240, 228)
        self.beige_border = "#D2C3AA"   # Color(210, 195, 170)
        self.beige_text = "#504132"     # Color(80, 65, 50)
        self.beige_button = "#C8B496"   # Color(200, 180, 150)
        self.beige_accent = "#AA8C64"   # Color(170, 140, 100)
        
        # Warna capsule/pill (sama persis dengan Java)
        self.capsule_white = "#FFFFFF"  # Color(255, 255, 255)
        self.capsule_border = "#DCDCDC" # Color(220, 220, 220)
        self.capsule_text = "#646464"   # Color(100, 100, 100)
        
        # Warna tombol (sama persis dengan Java)
        self.green_btn = "#4CAF50"      # Color(76, 175, 80)
        self.orange_btn = "#FF9800"     # Color(255, 152, 0)
        self.blue_btn = "#2196F3"       # Color(33, 150, 243)
        
        # Inisialisasi Text-to-Speech
        try:
            self.tts = pyttsx3.init()
            self.tts.setProperty('rate', 150)
            self.tts.setProperty('volume', 0.9)
            self.tts_available = True
        except Exception as e:
            print(f"TTS Error: {e}")
            self.tts_available = False
        
        self.setup_gui()
    
    def setup_gui(self):
        self.root.title("Sistem Antrian Bank")
        self.root.geometry("850x650")
        self.root.configure(bg=self.beige_bg)
        self.root.resizable(True, True)
        
        # Main panel (sama dengan BorderLayout)
        main_panel = tk.Frame(self.root, bg=self.beige_bg)
        main_panel.pack(fill=tk.BOTH, expand=True, padx=15, pady=15)
        
        # ==================== TOP PANEL ====================
        top_panel = tk.LabelFrame(
            main_panel,
            text="Ambil Antrian",
            font=('Segoe UI', 11, 'bold'),
            bg=self.beige_panel,
            fg=self.beige_accent,
            bd=2,
            relief=tk.GROOVE
        )
        top_panel.pack(fill=tk.X, pady=(0, 15))
        
        input_panel = tk.Frame(top_panel, bg=self.beige_panel)
        input_panel.pack(pady=15, padx=15)
        
        # Label Nama
        name_label = tk.Label(
            input_panel,
            text="Nama:",
            font=('Segoe UI', 12),
            bg=self.beige_panel,
            fg=self.beige_text
        )
        name_label.pack(side=tk.LEFT, padx=(0, 10))
        
        # TextField untuk input nama
        self.name_entry = tk.Entry(
            input_panel,
            font=('Segoe UI', 12),
            bg=self.beige_bg,
            fg=self.beige_text,
            relief=tk.SOLID,
            bd=1,
            width=25
        )
        self.name_entry.pack(side=tk.LEFT, padx=(0, 15))
        
        # Tombol capsule style
        take_button = self.create_capsule_button(
            input_panel, "Ambil Antrian", self.green_btn, "#FFFFFF"
        )
        take_button.pack(side=tk.LEFT, padx=5)
        
        call_button = self.create_capsule_button(
            input_panel, "Panggil Antrian", self.orange_btn, "#FFFFFF"
        )
        call_button.pack(side=tk.LEFT, padx=5)
        
        show_button = self.create_capsule_button(
            input_panel, "Tampilkan Antrian", self.blue_btn, "#FFFFFF"
        )
        show_button.pack(side=tk.LEFT, padx=5)
        
        # ==================== CALL PANEL ====================
        call_panel = tk.LabelFrame(
            main_panel,
            text="Antrian Dipanggil",
            font=('Segoe UI', 11, 'bold'),
            bg=self.beige_panel,
            fg=self.beige_accent,
            bd=2,
            relief=tk.GROOVE
        )
        call_panel.pack(fill=tk.X, pady=(0, 15))
        
        # Capsule label untuk antrian yang dipanggil
        self.called_label = self.create_capsule_label(call_panel, "-")
        self.called_label.pack(pady=15, padx=15)
        
        # ==================== LIST PANEL ====================
        list_panel = tk.LabelFrame(
            main_panel,
            text="Daftar Antrian",
            font=('Segoe UI', 11, 'bold'),
            bg=self.beige_panel,
            fg=self.beige_accent,
            bd=2,
            relief=tk.GROOVE
        )
        list_panel.pack(fill=tk.BOTH, expand=True)
        
        # Listbox untuk menampilkan antrian (mirip JList di Java)
        list_frame = tk.Frame(list_panel, bg=self.beige_panel)
        list_frame.pack(fill=tk.BOTH, expand=True, padx=10, pady=10)
        
        self.list_model = tk.Listbox(
            list_frame,
            font=('Segoe UI', 12),
            bg=self.beige_bg,
            fg=self.beige_text,
            selectbackground=self.beige_button,
            selectforeground="#FFFFFF",
            height=12,
            relief=tk.SOLID,
            bd=1
        )
        
        scrollbar = tk.Scrollbar(list_frame, orient=tk.VERTICAL, command=self.list_model.yview)
        self.list_model.configure(yscrollcommand=scrollbar.set)
        
        self.list_model.pack(side=tk.LEFT, fill=tk.BOTH, expand=True)
        scrollbar.pack(side=tk.RIGHT, fill=tk.Y)
        
        # ==================== STATUS BAR ====================
        self.status_label = tk.Label(
            self.root,
            text="Jumlah antrian: 0",
            font=('Segoe UI', 11),
            bg=self.beige_bg,
            fg=self.beige_text,
            relief=tk.SUNKEN,
            anchor=tk.CENTER,
            height=2
        )
        self.status_label.pack(side=tk.BOTTOM, fill=tk.X)
        
        # Event handlers
        take_button.config(command=self.take_queue)
        call_button.config(command=self.call_queue)
        show_button.config(command=self.show_queue)
        self.name_entry.bind('<Return>', lambda e: self.take_queue())
        
        # Tampilkan antrian awal
        self.show_queue()
    
    def create_capsule_button(self, parent, text, bg_color, fg_color):
        """Membuat tombol dengan efek capsule/pill"""
        button = tk.Button(
            parent,
            text=text,
            font=('Segoe UI', 11, 'bold'),
            bg=bg_color,
            fg=fg_color,
            relief=tk.RAISED,
            bd=1,
            padx=20,
            pady=8,
            cursor='hand2'
        )
        
        # Efek hover (mirip mouseEntered/mouseExited di Java)
        def on_enter(e):
            button.config(bg=self.darker_color(bg_color))
        
        def on_leave(e):
            button.config(bg=bg_color)
        
        button.bind("<Enter>", on_enter)
        button.bind("<Leave>", on_leave)
        
        return button
    
    def create_capsule_label(self, parent, text):
        """Membuat label dengan efek capsule/pill putih"""
        label = tk.Label(
            parent,
            text=text,
            font=('Segoe UI', 13, 'bold'),
            bg=self.capsule_white,
            fg=self.capsule_text,
            relief=tk.SOLID,
            bd=1,
            padx=30,
            pady=10,
            width=30
        )
        return label
    
    def darker_color(self, color):
        """Membuat warna lebih gelap (mirip darker() di Java)"""
        color = color.lstrip('#')
        r = int(color[0:2], 16)
        g = int(color[2:4], 16)
        b = int(color[4:6], 16)
        
        r = max(0, r - 30)
        g = max(0, g - 30)
        b = max(0, b - 30)
        
        return f'#{r:02x}{g:02x}{b:02x}'
    
    def speak(self, text):
        """Fitur panggilan suara (mirip speak() di Java)"""
        def _speak():
            try:
                os_name = platform.system().lower()
                
                if os_name == "darwin":  # Mac
                    import subprocess
                    subprocess.run(["say", text])
                elif os_name == "windows":  # Windows
                    if self.tts_available:
                        self.tts.say(text)
                        self.tts.runAndWait()
                elif os_name == "linux":  # Linux
                    import subprocess
                    subprocess.run(["espeak", text])
                print(f"[SUARA] {text}")
            except Exception as e:
                print(f"[SUARA GAGAL] {text} - {e}")
        
        threading.Thread(target=_speak, daemon=True).start()
    
    def take_queue(self):
        """Fitur ambil antrian (mirip takeQueue() di Java)"""
        name = self.name_entry.get().strip()
        
        if not name:
            messagebox.showwarning("Peringatan", "Masukkan nama terlebih dahulu!")
            return
        
        # Generate nomor antrian otomatis
        self.counter += 1
        customer = {'number': self.counter, 'name': name}
        self.queue.append(customer)
        
        # Tampilkan data dalam antrian
        self.show_queue()
        
        # Clear input
        self.name_entry.delete(0, tk.END)
        
        messagebox.showinfo(
            "Sukses",
            f"Antrian berhasil diambil!\nNomor: {customer['number']}\nNama: {customer['name']}"
        )
    
    def call_queue(self):
        """Fitur panggil antrian (mirip callQueue() di Java)"""
        if not self.queue:
            messagebox.showwarning("Peringatan", "Tidak ada antrian!")
            return
        
        # Ambil antrian pertama (mirip poll() di Java)
        customer = self.queue.popleft()
        
        # Menampilkan nomor dan nama yang dipanggil (dalam capsule)
        self.called_label.config(text=f"Nomor {customer['number']} - {customer['name']}")
        
        # Panggilan suara
        voice_text = f"Nomor antrian {customer['number']}, {customer['name']}, silakan menuju loket"
        self.speak(voice_text)
        
        # Tampilkan data antrian yang tersisa
        self.show_queue()
        
        messagebox.showinfo(
            "Panggil Antrian",
            f"Panggilan Antrian:\nNomor: {customer['number']}\nNama: {customer['name']}"
        )
    
    def show_queue(self):
        """Fitur tampilkan antrian (mirip showQueue() di Java)"""
        # Clear listbox
        self.list_model.delete(0, tk.END)
        
        # Tampilkan semua antrian
        if not self.queue:
            self.list_model.insert(tk.END, "Tidak ada antrian")
        else:
            for customer in self.queue:
                self.list_model.insert(tk.END, f"Nomor: {customer['number']} | Nama: {customer['name']}")
        
        # Update status bar
        self.status_label.config(text=f"Jumlah antrian: {len(self.queue)}")


if __name__ == "__main__":
    root = tk.Tk()
    app = QueueGUI(root)
    root.mainloop()

File: QueueGUI.java
import java.awt.*;
import java.util.LinkedList;
import java.util.Queue;
import javax.swing.*;

public class QueueGUI extends JFrame {
    
    // Queue menggunakan LinkedList
    private Queue<Customer> queue;
    private int counter;
    
    // Komponen GUI
    private JTextField nameEntry;
    private JList<String> queueList;
    private DefaultListModel<String> listModel;
    private JLabel calledLabel;
    private JLabel statusLabel;
    
    // Warna tema Beige
    private Color beigeBg = new Color(250, 245, 235);
    private Color beigePanel = new Color(245, 240, 228);
    private Color beigeBorder = new Color(210, 195, 170);
    private Color beigeText = new Color(80, 65, 50);
    private Color beigeButton = new Color(200, 180, 150);
    private Color beigeAccent = new Color(170, 140, 100);
    
    // Warna untuk capsule/pill
    private Color capsuleWhite = new Color(255, 255, 255);
    private Color capsuleBorder = new Color(220, 220, 220);
    private Color capsuleText = new Color(100, 100, 100);
    
    // Class Customer
    private static class Customer {
        int number;
        String name;
        
        Customer(int number, String name) {
            this.number = number;
            this.name = name;
        }
    }
    
    public QueueGUI() {
        // Inisialisasi Queue
        queue = new LinkedList<>();
        counter = 0;
        
        // Setup Frame
        setTitle("Sistem Antrian Bank");
        setSize(850, 650);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLayout(new BorderLayout());
        getContentPane().setBackground(beigeBg);
        
        createWidgets();
        setLocationRelativeTo(null);
    }
    
    // Fitur panggilan suara
    private void speak(String text) {
        new Thread(() -> {
            try {
                String os = System.getProperty("os.name").toLowerCase();
                
                if (os.contains("win")) {
                    String command = "powershell -Command \"Add-Type -AssemblyName System.Speech; " +
                                   "$synth = New-Object System.Speech.Synthesis.SpeechSynthesizer; " +
                                   "$synth.Speak('" + text + "');\"";
                    Runtime.getRuntime().exec(command);
                } else if (os.contains("mac")) {
                    Runtime.getRuntime().exec(new String[]{"say", text});
                } else if (os.contains("linux")) {
                    Runtime.getRuntime().exec(new String[]{"espeak", text});
                }
                System.out.println("[SUARA] " + text);
            } catch (Exception e) {
                System.out.println("[SUARA GAGAL] " + text);
            }
        }).start();
    }
    
    // Method untuk membuat capsule/pill button
    private JButton createCapsuleButton(String text, Color bgColor, Color textColor) {
        JButton button = new JButton(text) {
            @Override
            protected void paintComponent(Graphics g) {
                Graphics2D g2 = (Graphics2D) g.create();
                g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
                
                // Warna background
                if (getModel().isPressed()) {
                    g2.setColor(bgColor.darker());
                } else if (getModel().isRollover()) {
                    g2.setColor(bgColor.brighter());
                } else {
                    g2.setColor(bgColor);
                }
                
                // Gambar capsule (rounded rectangle)
                g2.fillRoundRect(0, 0, getWidth(), getHeight(), 30, 30);
                
                // Border capsule
                g2.setColor(capsuleBorder);
                g2.drawRoundRect(0, 0, getWidth() - 1, getHeight() - 1, 30, 30);
                
                // Teks
                g2.setColor(textColor);
                FontMetrics fm = g2.getFontMetrics();
                int x = (getWidth() - fm.stringWidth(getText())) / 2;
                int y = (getHeight() + fm.getAscent()) / 2 - 2;
                g2.drawString(getText(), x, y);
                
                g2.dispose();
            }
        };
        
        button.setFont(new Font("Segoe UI", Font.BOLD, 13));
        button.setForeground(textColor);
        button.setBackground(bgColor);
        button.setBorder(BorderFactory.createEmptyBorder(10, 25, 10, 25));
        button.setFocusPainted(false);
        button.setContentAreaFilled(false);
        button.setCursor(new Cursor(Cursor.HAND_CURSOR));
        button.setOpaque(false);
        
        return button;
    }
    
    // Method untuk membuat capsule label (untuk status)
    private JLabel createCapsuleLabel(String text) {
        JLabel label = new JLabel(text) {
            @Override
            protected void paintComponent(Graphics g) {
                Graphics2D g2 = (Graphics2D) g.create();
                g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
                
                // Background putih
                g2.setColor(capsuleWhite);
                g2.fillRoundRect(0, 0, getWidth(), getHeight(), 30, 30);
                
                // Border abu-abu
                g2.setColor(capsuleBorder);
                g2.drawRoundRect(0, 0, getWidth() - 1, getHeight() - 1, 30, 30);
                
                // Teks
                g2.setColor(capsuleText);
                FontMetrics fm = g2.getFontMetrics();
                int x = (getWidth() - fm.stringWidth(getText())) / 2;
                int y = (getHeight() + fm.getAscent()) / 2 - 2;
                g2.drawString(getText(), x, y);
                
                g2.dispose();
            }
        };
        
        label.setFont(new Font("Segoe UI", Font.BOLD, 14));
        label.setForeground(capsuleText);
        label.setBorder(BorderFactory.createEmptyBorder(8, 20, 8, 20));
        label.setOpaque(false);
        label.setHorizontalAlignment(SwingConstants.CENTER);
        
        return label;
    }
    
    private void createWidgets() {
        // Main panel
        JPanel mainPanel = new JPanel(new BorderLayout(10, 10));
        mainPanel.setBorder(BorderFactory.createEmptyBorder(15, 15, 15, 15));
        mainPanel.setBackground(beigeBg);
        
        // ==================== TOP PANEL ====================
        JPanel topPanel = new JPanel(new BorderLayout());
        topPanel.setBorder(BorderFactory.createTitledBorder(
            BorderFactory.createLineBorder(beigeBorder),
            "Ambil Antrian"
        ));
        topPanel.setBackground(beigePanel);
        
        JPanel inputPanel = new JPanel(new FlowLayout(FlowLayout.CENTER, 15, 15));
        inputPanel.setBackground(beigePanel);
        
        JLabel nameLabel = new JLabel("Nama:");
        nameLabel.setForeground(beigeText);
        nameLabel.setFont(new Font("Segoe UI", Font.PLAIN, 14));
        
        nameEntry = new JTextField(25);
        nameEntry.setFont(new Font("Segoe UI", Font.PLAIN, 14));
        nameEntry.setBackground(beigeBg);
        nameEntry.setForeground(beigeText);
        nameEntry.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(beigeBorder),
            BorderFactory.createEmptyBorder(8, 10, 8, 10)
        ));
        
        // Tombol dengan capsule style
        JButton takeButton = createCapsuleButton("Ambil Antrian", new Color(76, 175, 80), Color.WHITE);
        JButton callButton = createCapsuleButton("Panggil Antrian", new Color(255, 152, 0), Color.WHITE);
        JButton showButton = createCapsuleButton("Tampilkan Antrian", new Color(33, 150, 243), Color.WHITE);
        
        inputPanel.add(nameLabel);
        inputPanel.add(nameEntry);
        inputPanel.add(takeButton);
        inputPanel.add(callButton);
        inputPanel.add(showButton);
        
        topPanel.add(inputPanel, BorderLayout.CENTER);
        
        // ==================== CALL PANEL ====================
        JPanel callPanel = new JPanel();
        callPanel.setBorder(BorderFactory.createTitledBorder(
            BorderFactory.createLineBorder(beigeBorder),
            "Antrian Dipanggil"
        ));
        callPanel.setBackground(beigePanel);
        callPanel.setLayout(new FlowLayout(FlowLayout.CENTER));
        
        // Capsule label untuk antrian yang dipanggil
        calledLabel = createCapsuleLabel("-");
        callPanel.add(calledLabel);
        
        // ==================== LIST PANEL ====================
        JPanel listPanel = new JPanel(new BorderLayout());
        listPanel.setBorder(BorderFactory.createTitledBorder(
            BorderFactory.createLineBorder(beigeBorder),
            "Daftar Antrian"
        ));
        listPanel.setBackground(beigePanel);
        
        listModel = new DefaultListModel<>();
        queueList = new JList<>(listModel);
        queueList.setFont(new Font("Segoe UI", Font.PLAIN, 13));
        queueList.setBackground(beigeBg);
        queueList.setForeground(beigeText);
        queueList.setSelectionBackground(new Color(200, 180, 150));
        queueList.setSelectionForeground(Color.WHITE);
        queueList.setFixedCellHeight(35);
        
        JScrollPane scrollPane = new JScrollPane(queueList);
        scrollPane.getViewport().setBackground(beigeBg);
        scrollPane.setBorder(BorderFactory.createLineBorder(beigeBorder));
        listPanel.add(scrollPane, BorderLayout.CENTER);
        
        // ==================== STATUS BAR ====================
        statusLabel = new JLabel("Jumlah antrian: 0");
        statusLabel.setForeground(beigeText);
        statusLabel.setBackground(beigeBg);
        statusLabel.setFont(new Font("Segoe UI", Font.PLAIN, 12));
        statusLabel.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(beigeBorder),
            BorderFactory.createEmptyBorder(8, 10, 8, 10)
        ));
        statusLabel.setOpaque(true);
        
        // Add all to main panel
        mainPanel.add(topPanel, BorderLayout.NORTH);
        mainPanel.add(callPanel, BorderLayout.CENTER);
        mainPanel.add(listPanel, BorderLayout.SOUTH);
        
        add(mainPanel, BorderLayout.CENTER);
        add(statusLabel, BorderLayout.SOUTH);
        
        // Event handlers
        takeButton.addActionListener(e -> takeQueue());
        callButton.addActionListener(e -> callQueue());
        showButton.addActionListener(e -> showQueue());
        nameEntry.addActionListener(e -> takeQueue());
    }
    
    // Fitur ambil antrian
    private void takeQueue() {
        String name = nameEntry.getText().trim();
        
        if (name.isEmpty()) {
            JOptionPane.showMessageDialog(this, 
                "Masukkan nama terlebih dahulu!", 
                "Peringatan", 
                JOptionPane.WARNING_MESSAGE);
            return;
        }
        
        // Generate nomor antrian otomatis
        counter++;
        Customer customer = new Customer(counter, name);
        queue.offer(customer);
        
        // Tampilkan data dalam antrian
        showQueue();
        
        // Clear input
        nameEntry.setText("");
        
        JOptionPane.showMessageDialog(this,
            "Antrian berhasil diambil!\nNomor: " + customer.number + "\nNama: " + customer.name,
            "Sukses",
            JOptionPane.INFORMATION_MESSAGE);
    }
    
    // Fitur panggil antrian
    private void callQueue() {
        if (queue.isEmpty()) {
            JOptionPane.showMessageDialog(this, 
                "Tidak ada antrian!", 
                "Peringatan", 
                JOptionPane.WARNING_MESSAGE);
            return;
        }
        
        // Ambil antrian pertama
        Customer customer = queue.poll();
        
        // Menampilkan nomor dan nama yang dipanggil (dalam capsule)
        calledLabel.setText("Nomor " + customer.number + " - " + customer.name);
        
        // Panggilan suara
        String voiceText = "Nomor antrian " + customer.number + ", " + customer.name + ", silakan menuju loket";
        speak(voiceText);
        
        // Tampilkan data antrian yang tersisa
        showQueue();
        
        JOptionPane.showMessageDialog(this,
            "Panggilan Antrian:\nNomor: " + customer.number + "\nNama: " + customer.name,
            "Panggil Antrian",
            JOptionPane.INFORMATION_MESSAGE);
    }
    
    // Fitur tampilkan antrian
    private void showQueue() {
        // Clear listbox
        listModel.clear();
        
        // Tampilkan semua antrian
        if (queue.isEmpty()) {
            listModel.addElement("Tidak ada antrian");
        } else {
            for (Customer customer : queue) {
                listModel.addElement("Nomor: " + customer.number + " | Nama: " + customer.name);
            }
        }
        
        // Update status bar
        statusLabel.setText("Jumlah antrian: " + queue.size());
    }
    
    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> {
            new QueueGUI().setVisible(true);
        });
    }
}